src/sync/client.rs

```diff
@@ -14,76 +14,81 @@
 
 //! Sync client of ttrpc.
 
-use nix::sys::socket::*;
-use nix::unistd::close;
-use std::collections::HashMap;
+#[cfg(unix)]
 use std::os::unix::io::RawFd;
+
+use std::collections::HashMap;
 use std::sync::mpsc;
 use std::sync::{Arc, Mutex};
-use std::{io, thread};
+use std::{thread};
+use std::time::Duration;
 
-#[cfg(target_os = "macos")]
-use crate::common::set_fd_close_exec;
-use crate::common::{client_connect, SOCK_CLOEXEC};
 use crate::error::{Error, Result};
+use crate::sync::sys::{ClientConnection};
 use crate::proto::{Code, Codec, MessageHeader, Request, Response, MESSAGE_TYPE_RESPONSE};
 use crate::sync::channel::{read_message, write_message};
-use std::time::Duration;
+
+#[cfg(windows)]
+use super::sys::PipeConnection;
 
 type Sender = mpsc::Sender<(Vec<u8>, mpsc::SyncSender<Result<Vec<u8>>>)>;
 type Receiver = mpsc::Receiver<(Vec<u8>, mpsc::SyncSender<Result<Vec<u8>>>)>;
 
 /// A ttrpc Client (sync).
 #[derive(Clone)]
 pub struct Client {
-    _fd: RawFd,
+    _connection: Arc<ClientConnection>,
     sender_tx: Sender,
-    _client_close: Arc<ClientClose>,
 }
 
 impl Client {
     pub fn connect(sockaddr: &str) -> Result<Client> {
-        let fd = unsafe { client_connect(sockaddr)? };
-        Ok(Self::new(fd))
+        let conn = ClientConnection::client_connect(sockaddr)?;
+        
+        Self::new_client(conn)
     }
 
+    #[cfg(unix)]
     /// Initialize a new [`Client`] from raw file descriptor.
     pub fn new(fd: RawFd) -> Client {
-        let (sender_tx, rx): (Sender, Receiver) = mpsc::channel();
-
-        let (recver_fd, close_fd) =
-            socketpair(AddressFamily::Unix, SockType::Stream, None, SOCK_CLOEXEC).unwrap();
+        let conn = ClientConnection::new(fd);
 
-        // MacOS doesn't support descriptor creation with SOCK_CLOEXEC automically,
-        // so there is a chance of leak if fork + exec happens in between of these calls.
-        #[cfg(target_os = "macos")]
-        {
-            set_fd_close_exec(recver_fd).unwrap();
-            set_fd_close_exec(close_fd).unwrap();
-        }
-
-        let client_close = Arc::new(ClientClose { fd, close_fd });
+        Self::new_client(conn).unwrap_or_else(|e| {
+            panic!(
```

Comment:
it would be better if you can add some comments here like: TODO: upgrade the API of Client::new and remove this panic when the major version releasing.

Reply:
sure, is a comment enough or should I mark this a depreciated and add a new function like `pub fn new(fd: RawFd) -> Result<Client>`? I can do this in a follow up PR too if that makes sense.

Reply:
I think it's enough currently. the return value of the method could be changed in a new pr as you said, it's welcome and I will merge that pr just before I bump the major version(0.8.0).